File: Difficulty/A. Easy/1431-kids-with-the-greatest-number-of-candies/1431-kids-with-the-greatest-number-of-candies.py
from typing import List

class Solution:
    def kidsWithCandies(self, candies: List[int], extraCandies: int) -> List[bool]:
        """
        Determines which kids can have the greatest number of candies if given extraCandies.

        Parameters:
        candies (List[int]): A list where each element represents the number of candies a kid has.
        extraCandies (int): The number of extra candies that can be given to any one kid.

        Returns:
        List[bool]: A list of booleans indicating whether each kid can have the greatest or 
        equal number of candies if given the extraCandies.
        """
        # Find the maximum number of candies any kid currently has
        max_candies = max(candies)
        
        # List comprehension to build the result in a single pass
        return [(candie + extraCandies >= max_candies) for candie in candies]

File: Difficulty/A. Easy/1431-kids-with-the-greatest-number-of-candies/1431-kids-with-the-greatest-number-of-candies.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/**
 * @brief Determines if each kid can have the greatest number of candies
 *        after receiving the extra candies.
 *
 * @param candies (int*): Array representing number of candies each kid has.
 * @param candiesSize (int): Size of the candies array.
 * @param extraCandies (int): Number of extra candies available.
 * @param returnSize (int*): Pointer to return the size of the result array.
 *
 * @return (bool*): Boolean array where each element is true if the corresponding kid 
 *                  can have the greatest number of candies, otherwise false.
 */
bool* kidsWithCandies(int* candies, int candiesSize, int extraCandies, int* returnSize) {
    int maxCandies = 0;
    
    // Find the maximum number of candies
    for (int i = 0; i < candiesSize; i++) {
        if (candies[i] > maxCandies) {
            maxCandies = candies[i];
        }
    }

    // Allocate memory for the result array
    bool* result = (bool*)malloc(candiesSize * sizeof(bool));
    *returnSize = candiesSize;

    // Determine if each kid can have the greatest number of candies
    for (int i = 0; i < candiesSize; i++) {
        if (candies[i] + extraCandies >= maxCandies) {
            result[i] = true;
        } else {
            result[i] = false;
        }
    }

    return result;
}

